posts/Gunnar/2018-02-26-putting-bean-validation-constraints-to-multimaps.adoc

```diff
@@ -0,0 +1,217 @@
+
+== Registering the Value Extractors
+
+Having created the two value extractors for `Multimap`, they must be registered with the Bean Validation engine.
+There are multiple ways to do so (e.g. we could pass them to the bootstrap API when programmatically creating a `ValidatorFactory`),
+but the most convenient one is to rely on the service loader mechanism.
+
+For that we just need to declare a file named _META-INF/services/javax.validation.valueextraction.ValueExtractor_
+and give the fully-qualified names of our custom extractor implementations as the contents:
+
+[source]
+----
+com.example.MultimapKeyExtractor
+com.example.MultimapValueExtractor
+----
+
+The Bean Validation provider will automatically pick up all extractor implementations which are registered that way.
+
+Finally, let's run our example again and see how the resulting `ConstraintViolation` and its property path look like.
+(all the assertions in the example are true):
+
+[source,java]
+----
+Person bob = new Person();
+bob.emailsByType.put( "work", "bob@example.com" );
+bob.emailsByType.put( "work", "not-an-email" );
+
+Validator validator = Validation.buildDefaultValidatorFactory()
+    .getValidator();
+
+Set<ConstraintViolation<Bean>> violations = validator.validate (bean );
+assert violations.size() == 1;
+
+// one violation of the @Email constraint
+ConstraintViolation<Bean> violation = violations.iterator().next();
+assert violation.getInvalidValue().equals( "not-an-email" );
+assert violation.getConstraintDescriptor().getAnnotation().annotationType().equals( Email.class );
+
+Iterator<Node> pathNodes = violation.getPropertyPath().iterator();
+assert pathNodes.hasNext() == true;
+
+// first property path node
+Node node = pathNodes.next();
+assert node.getName().equals( "emailsByType" );
+assert node.getKind() == ElementKind.PROPERTY;
+
+assert pathNodes.hasNext() == true;
+
+// second node
+node = pathNodes.next();
+assert node.getName().equals( "<multimap value>" );
+assert node.getKind() == ElementKind.CONTAINER_ELEMENT;
+assert node.getKey().equals( "work" );
+
+assert pathNodes.hasNext() == false;
+----
+
+Of specific interest is the second node in the property path.
+It is of kind `CONTAINER_ELEMENT` and returns the name and key we passed in the value extractor.
+The invalid element's value can be obtained via `ConstraintViolation#getInvalidValue()`.
+
+== Summary
+
+While Bean Validation 2.0 comes with support for many container types http://beanvalidation.org/2.0/spec/#valueextractordefinition-builtinvalueextractors[out of the box]
+(besides the JDK collection types there's for instance also support for `Optional` and the property types from JavaFX),
+it also very easy to add support for other, custom container types by implementing the `ValueExtractor` SPI.
+
+To learn more, take a look at http://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#_resolution_algorithms[Value extraction] chapter of the Hibernate Validator reference guide.
```

Comment:
take a look at **the**?

Let's provide a HTTPS link and the anchor is wrong. The proper link should be: https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#chapter-valueextraction .